pkg/validate/validate_test.go

```diff
@@ -0,0 +1,304 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package validate
+
+import (
+	"bytes"
+	"strings"
+	"testing"
+
+	"github.com/ghodss/yaml"
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/kylelemons/godebug/diff"
+	"istio.io/operator/pkg/apis/istio/v1alpha2"
+	"istio.io/operator/pkg/util"
+)
+
+func TestUnmarshalKubernetes(t *testing.T) {
+	tests := []struct {
+		desc    string
+		yamlStr string
+		want    string
+	}{
+		{
+			desc:    "nil success",
+			yamlStr: "",
+			want:    "{}",
+		},
+		{
+			desc: "hpaSpec",
+			yamlStr: `
+hpaSpec:
+  maxReplicas: 10
+  minReplicas: 1
+  targetCPUUtilizationPercentage: 80
+  scaleTargetRef:
+    apiVersion: apps/v1
+    kind: Deployment
+    name: php-apache
+`,
+		},
+		{
+			desc: "resources",
+			yamlStr: `
+resources:
+  limits:
+    cpu: 444m
+    memory: 333Mi
+  requests:
+    cpu: 222m
+    memory: 111Mi
+`,
+		},
+		{
+			desc: "podDisruptionBudget",
+			yamlStr: `
+podDisruptionBudget:
+  maxUnavailable: 1
+  selector:
+    matchLabels:
+      app: pilot
+`,
+		},
+		{
+			desc: "readinessProbe",
+			yamlStr: `
+readinessProbe:
+  failureThreshold: 44
+  initialDelaySeconds: 11
+  periodSeconds: 22
+  successThreshold: 33
+  handler: {}
+`,
+		},
+		{
+			desc: "affinity",
+			yamlStr: `
+affinity:
+  podAffinity:
+    requiredDuringSchedulingIgnoredDuringExecution:
+    - labelSelector:
+        matchExpressions:
+        - key: security
+          operator: In
+          values:
+          - S1
+      topologyKey: failure-domain.beta.kubernetes.io/zone
+  podAntiAffinity:
+    preferredDuringSchedulingIgnoredDuringExecution:
+    - weight: 100
+      podAffinityTerm:
+        labelSelector:
+          matchExpressions:
+          - key: security
+            operator: In
+            values:
+            - S2
+        topologyKey: failure-domain.beta.kubernetes.io/zone
+`,
+		},
+		{
+			desc: "k8sObjectOverlay",
+			yamlStr: `
+overlays:
+- kind: Deployment
+  name: istio-citadel
+  patches:
+  - path: spec.template.spec.containers.name:galley.ports.containerPort:15014
+    value: 12345
+`,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			tk := &v1alpha2.TestKube{}
+			err := unmarshalWithJSONPB(tt.yamlStr, tk)
```

Comment:
this test failed for me with:
```
--- FAIL: TestUnmarshalKubernetes (0.00s)
    --- FAIL: TestUnmarshalKubernetes/k8sObjectOverlay (0.00s)
        validate_test.go:130: unmarshalWithJSONPB(k8sObjectOverlay): got error unknown field "value" in v1alpha2.K8SObjectOverlay_PathValue
```
note: I've rebased this branch on #34 , is it probably because of that?

Reply:
Hard to say - but I won't check this in until tests are passing. Most likely I'll have to cherry pick a few other changes from other PRs to make it all work.